src/main/java/org/truffleruby/core/time/TimeNodes.java

```diff
@@ -420,24 +421,47 @@ protected Object timeZone(RubyTime time) {
         protected RubyString timeStrftime(VirtualFrame frame, RubyTime time, Object format,
                 @CachedLibrary(limit = "2") RubyStringLibrary libFormat,
                 @Cached("libFormat.getRope(format)") Rope cachedFormat,
-                @Cached("compilePattern(cachedFormat)") List<Token> pattern,
-                @Cached RopeNodes.EqualNode equalNode) {
-            return makeStringNode.fromBuilderUnsafe(formatTime(time, pattern), CodeRange.CR_UNKNOWN);
+                @Cached(value = "compilePattern(cachedFormat)", dimensions = 1) Token[] pattern,
+                @Cached RopeNodes.EqualNode equalNode,
+                @Cached("formatToRopeBuilderCanBeFast(pattern)") boolean canUseFast) {
+            if (canUseFast) {
+                return makeStringNode.fromRope(formatTimeFast(time, pattern));
+            } else {
+                return makeStringNode.fromBuilderUnsafe(formatTime(time, pattern), CodeRange.CR_UNKNOWN);
+            }
+        }
+
+        protected boolean formatToRopeBuilderCanBeFast(Token[] pattern) {
+            return RubyDateFormatter.formatToRopeBuilderCanBeFast(pattern);
         }
 
         @Specialization(guards = "libFormat.isRubyString(format)")
         protected RubyString timeStrftime(VirtualFrame frame, RubyTime time, Object format,
                 @CachedLibrary(limit = "2") RubyStringLibrary libFormat) {
-            final List<Token> pattern = compilePattern(libFormat.getRope(format));
-            return makeStringNode.fromBuilderUnsafe(formatTime(time, pattern), CodeRange.CR_UNKNOWN);
+            return makeStringNode.fromBuilderUnsafe(
+                    formatTime(time, compilePattern(libFormat.getRope(format))),
+                    CodeRange.CR_UNKNOWN);
         }
 
         @TruffleBoundary
-        protected List<Token> compilePattern(Rope format) {
-            return RubyDateFormatter.compilePattern(format, false, getContext(), this);
+        protected Token[] compilePattern(Rope format) {
+            final List<Token> tokens = RubyDateFormatter.compilePattern(format, false, getContext(), this);
+            return tokens.toArray(RubyDateFormatter.EMPTY_TOKEN_ARRAY);
```

Comment:
Could you make RubyDateFormatter.compilePattern return a `Token[]` instead? (consistency)

Reply:
Done.